src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt

```diff
@@ -340,50 +339,51 @@ bool emberAfDiscoverCommandsReceivedResponseCallback(ClusterId clusterId, uint16
 {{#if isList}}
 void {{asUpperCamelCase parent.name}}Cluster{{asUpperCamelCase name}}ListAttributeFilter(TLV::TLVReader * tlvData, Callback::Cancelable * onSuccessCallback, Callback::Cancelable * onFailureCallback)
 {
-    // TODO: Add actual support for array and lists.
-    const uint8_t * message = nullptr;
-    uint16_t messageLen     = 0;
-    EmberAfStatus res = PrepareListFromTLV(tlvData, message, messageLen);
-    if (res != EMBER_ZCL_STATUS_SUCCESS)
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    size_t count = 0;
+    TLV::TLVType tmpType;
+    TLV::TLVReader reader;
+
+    if (tlvData != nullptr && tlvData->GetType() == chip::TLV::TLVType::kTLVType_Array)
+    {
+        reader.Init(*tlvData);
+        reader.EnterContainer(tmpType);
+    }
+    if (tlvData == nullptr || tlvData->GetType() != chip::TLV::TLVType::kTLVType_Array ||
+        chip::TLV::Utilities::Count(reader, count, false /* recursive */) != CHIP_NO_ERROR || !chip::CanCastTo<uint16_t>(count))
     {
         Callback::Callback<DefaultFailureCallback> * cb = Callback::Callback<DefaultFailureCallback>::FromCancelable(onFailureCallback);
-        cb->mCall(cb->mContext, res);
+        cb->mCall(cb->mContext, EMBER_ZCL_STATUS_INVALID_VALUE);
         return;
     }
-
-    CHECK_MESSAGE_LENGTH_VOID(2);
-    uint16_t count = Encoding::LittleEndian::Read16(message);
     {{chipType}} data[count];
-    for (size_t i = 0; i < count; i++)
+    for (size_t i = 0; i < count && CHIP_NO_ERROR == (err = reader.Next()); i++)
     {
         {{#if isStruct}}
-        {{#chip_attribute_list_entryTypes}}
-        {{#if (isString type)}}
-        CHECK_STATUS_VOID(ReadByteSpan(message, {{size}}, &data[i].{{name}}));
-        messageLen = static_cast<uint16_t>(messageLen - {{size}});
-        {{else}}
-        CHECK_MESSAGE_LENGTH_VOID({{size}});
-        data[i].{{name}} = emberAfGet{{asReadType type}}(message, 0, {{size}});
-        {{/if}}
-        message += {{size}};
-        {{/chip_attribute_list_entryTypes}}
+        SuccessOrExit(reader.Get(data[i]));
         {{else}}
         {{#if (isString type)}}
-        CHECK_STATUS_VOID(ReadByteSpan(message, messageLen, &data[i]));
-        uint16_t entryLength = static_cast<uint16_t>(data[i].size() + kByteSpanSizeLengthInBytes);
-        messageLen = static_cast<uint16_t>(messageLen - entryLength);
-        message += entryLength;
+        const uint8_t * dataPtr = nullptr;
+        SuccessOrExit(reader.GetDataPtr(dataPtr));
+        data[i] = chip::ByteSpan(dataPtr, reader.GetLength());
         {{else}}
-        CHECK_MESSAGE_LENGTH_VOID({{size}});
-        data[i] = emberAfGet{{asReadType type}}(message, 0, {{size}});
-        message += {{size}};
+        SuccessOrExit(reader.Get(data[i]));
```

Comment:
Instead of
 ```
{{#if isStruct}}
SuccessOrExit(reader.Get(data[i]));
{{else}}
{{#if (isString type)}}
const uint8_t * dataPtr = nullptr;
SuccessOrExit(reader.GetDataPtr(dataPtr));
data[i] = chip::ByteSpan(dataPtr, reader.GetLength());
{{else}}
SuccessOrExit(reader.Get(data[i]));
{{/if}}
{{/if}}
```

you may just use:  ```SuccessOrExit(reader.Get(data[i]));```